clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiDriver.java

```diff
@@ -0,0 +1,464 @@
+        KafkaFutureImpl<V> future = futures.get(key);
+        if (future == null) {
+            log.warn("Attempt to complete future for {}, which was not requested", key);
+        } else {
+            clear(key);
+            future.complete(value);
+        }
+    }
+
+    /**
+     * Check whether any requests need to be sent. This should be called immediately
+     * after the driver is constructed and then again after each request returns
+     * (i.e. after {@link #onFailure(long, RequestSpec, Throwable)} or
+     * {@link #onResponse(long, RequestSpec, AbstractResponse)}).
+     *
+     * @return A list of requests that need to be sent
+     */
+    public List<RequestSpec<K>> poll() {
+        List<RequestSpec<K>> requests = new ArrayList<>();
+        collectLookupRequests(requests);
+        collectFulfillmentRequests(requests);
+        return requests;
+    }
+
+    /**
+     * Get a map of the futures that are awaiting completion.
+     */
+    public Map<K, KafkaFutureImpl<V>> futures() {
+        return futures;
+    }
+
+    /**
+     * Callback that is invoked when a `Call` returns a response successfully.
+     */
+    public void onResponse(
+        long currentTimeMs,
+        RequestSpec<K> spec,
+        AbstractResponse response
+    ) {
+        clearInflightRequest(currentTimeMs, spec);
+        if (spec.scope instanceof AdminApiDriver.BrokerScope) {
+            int brokerId = ((BrokerScope) spec.scope).destinationBrokerId;
+            AdminApiHandler.ApiResult<K, V> result = handler.handleResponse(
+                brokerId,
+                spec.keys,
+                response
+            );
+            result.completedKeys.forEach(this::complete);
+            result.failedKeys.forEach(this::completeExceptionally);
+            result.unmappedKeys.forEach(this::unmap);
+        } else {
+            DynamicKeyMapping<K> mapping = dynamicMapping.orElseThrow(() ->
+                new IllegalStateException("Received response for unexpected dynamic lookup scope " + spec.scope)
+            );
+
+            AdminApiLookupStrategy.LookupResult<K> result = mapping.lookupStrategy.handleResponse(
+                spec.keys,
+                response
+            );
+            result.failedKeys.forEach(this::completeExceptionally);
+            result.mappedKeys.forEach(this::map);
+        }
+    }
+
+    /**
+     * Callback that is invoked when a `Call` is failed.
+     */
+    public void onFailure(
+        long currentTimeMs,
+        RequestSpec<K> spec,
+        Throwable t
+    ) {
+        clearInflightRequest(currentTimeMs, spec);
+        if (t instanceof DisconnectException) {
+            log.debug("Node disconnected before response could be received for request {}. " +
+                "Will attempt retry", spec.request);
+            dynamicMapping.ifPresent(mapping -> {
+                if (mapping.keys.containsAll(spec.keys)) {
+                    spec.keys.forEach(this::unmap);
```

Comment:
Is this required if this error happens when sending fulfillment request?

Reply:
Let me add a comment here. The idea is that we want to re-attempt discovery of the brokerId mapping. Say, for example, that we found the leader of partition 0 is broker A. Before we connect to it, broker A fails, which results in a disconnect. Rather than continuing to retry on broker A until the timeout expires, we want to allow an opportunity to to find the new leader. The test case `testRetryLookupAfterDisconnect` shows the basic idea.

Reply:
Thanks for explanation. Is the test case correspond to production code? The exception passed to `onFailure` is wrapped to `TimeoutException` (see https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L821). not sure whether `t instanceof DisconnectException` work wells here.